### PR TITLE
[SelectionDAG] Stop storing EVTs in a function scoped static std::set.

### LINKED ISSUE #88233: llvm:SelectionDAG: memory leak caused by static std::set EVTs
Hi!

We're using LLVM for jit compilation in our in-memory database, therefore memory handling is crucial. While switching from llvm 13 to 18 we saw that the std::set EVTs causes a memory leak in SDNode::getValueTypeList(). As far as I understand the code, for each extended EVT (we're using i96) a new value is inserted to the set if the pointer of the Type is different. An this does happen for identical types, if the compilation is done with a new instance of the llvm classes.

A possible fix for this leak would be to make EVTs a member of SelectionDAG. However, SDNode::getValueTypeList() has no access to the SelectionDAG instance, and passing EVTs as parameter would require many further changes, since getValueTypeList is called from SDNode::getSDVTList() and this is called in many locations, also outside of SelectionDAG.

What can we do to close this memory leak?

@nhaehnle, you switched EVTs from a managed static to a local static. Do you know how to handle the situation in a better way?

### PR DESCRIPTION
**Pull request by topperc**
EVTs potentially contain a Type * that points into memory owned by an LLVMContext. Storing them in a function scoped static means they may outlive the LLVMContext they point to.

This std::set is used to unique single element VT lists containing a single extended EVT. Single element VT list with a simple EVT are uniqued by a separate cache indexed by the MVT::SimpleValueType enum. VT lists with more than one element are uniqued by a FoldingSet owned by the SelectionDAG object.

This patch moves the single element cache into SelectionDAG so that it will be destroyed when SelectionDAG is destroyed.

Fixes #88233

### CODE REVIEW

**Comment by efriedma-quic** [comment]
Is there some reason you're switching to FoldingSet, instead of just sticking with std::set?  Not sure what the performance of FoldingSet is like, but any performance difference is probably related to that switch, not whether the std::set is a global.

**Comment by topperc** [comment]
> Is there some reason you're switching to FoldingSet, instead of just sticking with std::set? Not sure what the performance of FoldingSet is like, but any performance difference is probably related to that switch, not whether the std::set is a global.

Only because the FoldingSet already existed for the multiple element case so it was easy to reuse. I'll try with a std::set.

**Comment by topperc** [comment]
> > Is there some reason you're switching to FoldingSet, instead of just sticking with std::set? Not sure what the performance of FoldingSet is like, but any performance difference is probably related to that switch, not whether the std::set is a global.
> 
> Only because the FoldingSet already existed for the multiple element case so it was easy to reuse. I'll try with a std::set.

Looks like moving the std::set into SelectionDAG gives better results. https://llvm-compile-time-tracker.com/compare.php?from=2fea1ccb6221238674562533684c51b63de248d4&to=76f11748278f882b13ed351c166797511e0ae825&stat=instructions:u

File: llvm/include/llvm/CodeGen/SelectionDAG.h [hunk 1]
```diff
@@ -247,6 +248,9 @@ class SelectionDAG {
   BlockFrequencyInfo *BFI = nullptr;
   MachineModuleInfo *MMI = nullptr;
 
+  /// Extended EVTs used for single value VTLists.
+  std::set<EVT, EVT::compareRawBits> EVTs;
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/SelectionDAG.h:252]: Did you try all the set types? Does DenseSet do any better?

**Review by arsenm — APPROVED** [review]

File: llvm/include/llvm/CodeGen/SelectionDAG.h (hunk 1 above)
Comment (topperc) [reply llvm/include/llvm/CodeGen/SelectionDAG.h:252]: I didn't try DenseSEt. Looks like we would need to add a DenseMapInfo for EVT including defining Empty and Tombstone values?

Comment (arsenm) [reply llvm/include/llvm/CodeGen/SelectionDAG.h:252]: Yes

Comment (efriedma-quic) [reply llvm/include/llvm/CodeGen/SelectionDAG.h:252]: The usage here depends on the address stability of the map entries (the VTList points at the map's key), and DenseSet doesn't provide that.  So you'd need define a struct like `EVTSetKey Key { EVT* Ptr; };`, define DenseMapInfo for that, then use `DenseSet<EVTSetKey>`.

I guess you could do that, but it seems like overkill.

**Pull request merged**